src/session_state.rs

```diff
@@ -2,26 +2,45 @@
 ///
 ///
 use embedded_nal::IpAddr;
+use embedded_time::{duration::Seconds, Clock, Instant};
 use heapless::{consts, String, Vec};
 
-pub struct SessionState {
+pub struct SessionState<C>
+where
+    C: Clock,
+{
     pub connected: bool,
-    pub keep_alive_interval: u16,
+
+    /// (embedded_time doesn't support u16.)
+    pub keep_alive_interval: Seconds<u32>,
+
+    /// Timestamp of the last message sent to the broker, for keep-alive generation.
+    pub last_write_time: Option<Instant<C>>,
+
+    /// If we sent a ping request without having received a response, stores the time
+    /// the request was sent.
+    pub pending_pingreq_time: Option<Instant<C>>,
```

Comment:
I think this would be a lot simpler if we just kept two state variables:
1. `pending_response: bool`
2. `next_ping`: Instant<C>`

Then if we hit `next_ping` while `pending_response` is set, we disconnect from the broker. This equates to setting the ping timeout equal to the ping rate.